### PR TITLE
[p2p] add a timed blocklist

### PR DESCRIPTION
**Pull request by LeoHChen**
blocklist peers not in the current committee for 120 seconds.

there is a risk that internal nodes may still
use the keys that are not in the current committee
if we increase the slot from 480 to 640.

Signed-off-by: Leo Chen <leo@harmony.one>

### CODE REVIEW

File: node/node.go [hunk 1]
```diff
@@ -561,7 +561,15 @@ func (node *Node) Start() error {
 					)
 
 					if err != nil {
-						errChan <- withError{err, msg.GetFrom()}
+						// block the peer if it is not in the current committee
+						// FIXME: this is for testing only, will need to enable it after we increase the slot to 640
+						// as the internal nodes may contain invalid keys and we don't want to block internal nodes
+						if err == shard.ErrValidNotInCommittee {
```
Comment (JackyWYX) [inline node/node.go:567]: Why are we only blocking this specific error? Maybe we should also block invalid signature and more?

**Comment by JackyWYX** [comment]
I see the risk of committee change which might even cause system break in that specific epoch change. But there is also risks for any epoch change since it could happen that new validators is not acknowledged by other validators. I would recommend also to test the epoch change scenario as well.

**Comment by JackyWYX** [comment]
Also some comments on previous PRs:

1. node/node.go:555: If node is not consensus bound, sould return `ValidationAccept`?
2. node/node.go:541: If node is delivering empty message like spamming, should we return `ValidationReject` instead?

**Comment by LeoHChen** [comment]
> Also some comments on previous PRs:
> 
> 1. node/node.go:555: If node is not consensus bound, sould return `ValidationAccept`?
> 2. node/node.go:541: If node is delivering empty message like spamming, should we return `ValidationReject` instead?
> 3. Shall we verify signature in p2p message validation before entering consensus?

1) this is not a risk as we already switch case the messageCategory before the validation. https://github.com/harmony-one/harmony/blob/14e017defc18d712a854f70104e173e6eea9c058/node/node.go#L547

2) agree, we shall probably reject empty message as well

3) verify the signature of p2p message is very expensive, also, as I raised the question of the duplicated signature verification in https://github.com/harmony-one/harmony/issues/3225,  I think we shall not do signature verification at p2p layer.

**Comment by LeoHChen** [comment]
> I see the risk of committee change which might even cause system break in that specific epoch change. But there is also risks for any epoch change since it could happen that new validators is not acknowledged by other validators. I would recommend also to test the epoch change scenario as well.

definitely will test on the epoch change scenario. For epoch change, there may be a risk, when the leader already moved to a new epoch, and validators still in the previous epoch, and keep sending messages signed with old keys, thus may be banned in the network for some time.  However, unless all nodes are doing the same kind of blskey manipulation, this shouldn't break the network.

Also, the block is timed, we can change it like 1 minute or so to shorter the block time.

File: node/node.go (hunk 1 above)
Comment (LeoHChen) [reply node/node.go:567]: Yes, we can also block other errors.  This is the initial intention as we have seen many error messages like this. Ofc, to protect the network, we shall also block on other errors.

**Comment by JackyWYX** [comment]
> verify the signature of p2p message is very expensive, also, as I raised the question of the duplicated signature verification in #3225, I think we shall not do signature verification at p2p layer.

I see your point. But I wonder whether we have done any spamming where attackers send consensus message with public key in committee but invalid signature?

**Comment by rlan35** [comment]
> > I see the risk of committee change which might even cause system break in that specific epoch change. But there is also risks for any epoch change since it could happen that new validators is not acknowledged by other validators. I would recommend also to test the epoch change scenario as well.
> 
> definitely will test on the epoch change scenario. For epoch change, there may be a risk, when the leader already moved to a new epoch, and validators still in the previous epoch, and keep sending messages signed with old keys, thus may be banned in the network for some time. However, unless all nodes are doing the same kind of blskey manipulation, this shouldn't break the network.
> 
> Also, the block is timed, we can change it like 1 minute or so to shorter the block time.

I will definitely cause trouble during epoch transition. We should change the logic to accept both new epoch and last epoch's keys. More optimized way is to accept last epoch's key within a certain range of time after entering new epoch. But I think simply accept new and last epoch should be fine.

**Pull request closed without merging**